opentelemetry tracer: implement TraceIDRatioBased and ParentBased samplers

therealak12 opened the pull request:
Commit Message: implement TraceIDRatioBased and ParentBased samplers

Additional Description:
[A previous issue](https://github.com/envoyproxy/envoy/issues/35016) existed for adding opentelemetry samplers to Envoy so I didn't create a new one. The PR at hand implements two of these samplers.
The implementation and testing strategies are inspired by the OTEL's [Go](https://github.com/open-telemetry/opentelemetry-go) and [CPP](https://github.com/open-telemetry/opentelemetry-cpp) SDKs.

Risk Level: Low

Testing: The PR has unit and integration tests which I've run locally.

Docs Changes:
Release Notes:
Platform Specific Features:

therealak12 commented:
/retest

therealak12 commented:
/retest

therealak12 commented:
The `ci/do_ci.sh` coverage job is failing. The only unusual log entry I can find is a warning message. I'd appreciate any assistance in identifying and resolving the issue.

Here's the warning message I said about:
![Screenshot from 2024-12-21 19-13-02](https://github.com/user-attachments/assets/9efc55d7-12d9-47f0-ab04-0f436d651c87)

phlax commented:
the warning is not relevant

this is the important line

```console
Code coverage for source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based is lower than limit of 96.6 (84.8)
```

/coverage

therealak12 commented:
Thanks for your guidance, Dear @phlax. Tests are passing now.

wbpcode commented:
Thanks for the contribution. Please read the https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md to get more infomation about adding new extensions to see if the new extensions meet the requirement or not.

Thanks again!

therealak12 commented:
> Please read the https://github.com/envoyproxy/envoy/blob/main/CONTRIBUTING.md and https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md to get more information about adding new extensions to see if the new extensions meet the requirement or not.

Thanks for the links, @wbpcode! I’ve reviewed them and made the necessary adjustments to my PR. However, just to clarify, my PR doesn't add a new extension; rather, it implements two missing samplers for the OpenTelemetry tracer extension.

Could you please let me know if that matches your expectations?

therealak12 commented:
If you get a chance, I'd appreciate any feedback or comments when you have time. No rush, I understand you're busy.

wbpcode commented:
> Thanks for the links, @wbpcode! I’ve reviewed them and made the necessary adjustments to my PR. However, just to clarify, my PR doesn't add a new extension; rather, it implements two missing samplers for the OpenTelemetry tracer extension.
> Could you please let me know if that matches your expectations?

Sorry for the delay. I skipped messages from community these two days.

It's no doubt this add some new extensions. Although these extensions are not top level extensions but sub extensions for opentelmetry. But it doesn't change to much the characteristics.

So, at least, we should ensure the current owners of opentelmetry @alexanderellis @yanavlasov are willing to maintain these new two extensions.

wbpcode commented:
/wait-any

therealak12 commented:
I would appreciate it if you took a look at this when you have a moment.
@AlexanderEllis @yanavlasov

therealak12 commented:
Have you had a chance to review this PR? It's crucial for our tracing implementation.
@AlexanderEllis  @yanavlasov

therealak12 commented:
This change would be beneficial in our environment, where we handle a large volume of spans. It allows us to leverage the parent span's sampling decision. I would greatly appreciate it if you could review this PR.
@AlexanderEllis 
@yanavlasov

therealak12 commented:
"Dear @AlexanderEllis @yanavlasov, I’m still awaiting your review. In my opinion, this should have already been implemented, as it is a necessary feature rather than just a nice-to-have. Looking forward to your thoughts!"

therealak12 commented:
Please review😭

therealak12 commented:
Would it be possible for someone else to review this PR?
It seems like the current owners might be a bit tied up, and I’d really appreciate it if someone could take a look when they have a moment.

therealak12 commented:
No worries, @phlax.
All set. rebased on main and updated the Bazel deps.

phlax commented:
/lgtm deps (not really deps changes as such)

@wbpcode i think you will have to take a view on this

@therealak12 i think the blocker is finding a maintainer/codeowner to approve/agree to maintaining the code

therealak12 commented:
> @therealak12 i think the blocker is finding a maintainer/codeowner to approve/agree to maintaining the code

Could you please take a look at this PR? @yanavlasov

mschem-amz commented:
Hello all, is there any work that needs to be done to get this merged? I'm happy to take look if so. This PR is very valuable to my team.

ravenblackx commented:
Once again nagging both @yanavlasov and @wbpcode for either of them to take a look.

wbpcode commented:
@mschem-amz sorry for the delay. I can help with review, but I am not the extension owner, I don't know whether this meet our contribution policy. cc @yanavlasov

wbpcode commented on api/envoy/extensions/tracers/opentelemetry/samplers/v3/trace_id_ratio_based_sampler.proto:
```diff
@@ -0,0 +1,29 @@
+  // return RECORD_AND_SAMPLE.
+  // required
+  // [#extension-category: envoy.tracers.opentelemetry.samplers]
+  double ratio = 1;
```
nit: may use `FractionalPercent`

wbpcode replied:
To keep the consistance with our exist API. Please use `type.v3.Percent` or type.v3. FractionalPercent`

therealak12 replied:
I've done that. Could you please verify if it meets your expectations?

wbpcode reviewed (commented):
Thanks for the contribution and sorry for the delay. LGTM to the code and API with some some comments. 

Although I am not owner, but seems this is waiting for a while, I will approve and merge this if no owner object this explicity.

therealak12 commented on source/extensions/tracers/opentelemetry/samplers/parent_based/config.cc:
```diff
@@ -0,0 +1,38 @@
+      const envoy::extensions::tracers::opentelemetry::samplers::v3::ParentBasedSamplerConfig&>(
+      *mptr, context.messageValidationVisitor());
+  auto* factory =
+      Envoy::Config::Utility::getFactory<SamplerFactory>(proto_config.wrapped_sampler());
```
I've applied your suggestion. The CI is likely failing due to an error not related to this change.

therealak12 commented on source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based/trace_id_ratio_based_sampler.cc:
```diff
@@ -0,0 +1,113 @@
+ *
+ * Adapted from https://github.com/open-telemetry/opentelemetry-cpp
+ */
+uint64_t ratioToThreshold(double ratio) noexcept {
+  if (ratio <= 0.0) {
+    return 0;
+  }
+  if (ratio >= 1.0) {
+    return UINT64_MAX;
+  }
+
+  // We can't directly return ratio * UINT64_MAX because of precision issues with doubles.
+  //
+  // UINT64_MAX is (2^64 - 1), but a double has limited precision and rounds up when
+  // converting large numbers. For probabilities >= 1 - (2^-54), multiplying by UINT64_MAX
+  // can cause overflow and the result wraps around to zero!
+  // We know (UINT64_MAX)*ratio = (UINT32_MAX*2^32 + UINT32_MAX)*ratio. So, To avoid this,
+  // we calculate the high and low 32 bits separately by using the double's precision to
+  // get the most accurate result.
+  // Then we add the the high and low bits to get the result.
+  const double product = UINT32_MAX * ratio;
+  double hi_bits;
+  double lo_bits = modf(product, &hi_bits);
+  lo_bits = ldexp(lo_bits, 32) + product;
+  return (static_cast<uint64_t>(hi_bits) << 32) + static_cast<uint64_t>(lo_bits);
+}
```
I'll apply the suggestion and let you know when I'm done.

therealak12 replied:
I’ve made some edits.

could you check if they match your intent?

ravenblackx replied:
Ping @wbpcode

therealak12 replied:
Is there anything I could do to help expedite the review process?

wbpcode commented on source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based/trace_id_ratio_based_sampler.cc:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Converts a ratio in [0, 1] to a threshold in [0, MILLION].
+ */
+uint64_t ratioToThreshold(double ratio) noexcept {
+  const uint64_t MAX_VALUE = Envoy::ProtobufPercentHelper::fractionalPercentDenominatorToInt(
+      envoy::type::v3::FractionalPercent::MILLION);
+
+  if (ratio <= 0.0) {
+    return 0;
+  }
+  if (ratio >= 1.0) {
+    return MAX_VALUE;
+  }
+
+  return static_cast<uint64_t>(ratio * static_cast<double>(MAX_VALUE));
+}
```
I actually didn't get this. Although `double` is find as ratio for me. But if we use fractionl percent helper here, why not use the `envoy.type.v3.FractionalPercent` directly and use `evaluateFractionalPercent` to get the decsion?

therealak12 replied:
Hi @wbpcode,
would you mind taking another look at the code?

wbpcode commented on api/envoy/extensions/tracers/opentelemetry/samplers/v3/trace_id_ratio_based_sampler.proto:
```diff
@@ -20,10 +22,9 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
   // required
   // [#extension-category: envoy.tracers.opentelemetry.samplers]
-  double ratio = 1;
+  type.v3.FractionalPercent sampling_percentage = 2;
```
```suggestion
  type.v3.FractionalPercent sampling_percentage = 1;
```

wbpcode reviewed (changes requested):
Sorry, I forget a comment.

therealak12 commented on source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based/trace_id_ratio_based_sampler.cc:
```diff
@@ -0,0 +1,84 @@
+  Envoy::safeMemcpyUnsafeSrc(&first_8_bytes, buffer);
+
+  return first_8_bytes;
+}
```
There were no explicit checks previously, and I assumed the trace ID would always be valid, which, of course, was an unsafe assumption.

The necessary check has now been added.

wbpcode commented on source/extensions/tracers/opentelemetry/samplers/trace_id_ratio_based/BUILD:
```diff
@@ -26,6 +26,8 @@ envoy_cc_library(
     srcs = ["trace_id_ratio_based_sampler.cc"],
     hdrs = ["trace_id_ratio_based_sampler.h"],
     deps = [
+        # "//source/common/common:minimal_logger_lib",
+        "//source/common/common:logger_lib",
```
```suggestion
        "//source/common/common:logger_lib",
```

wbpcode reviewed (commented):
LGTM with one nit comment. I think we can land this today or tomorrow. :) Thanks again for all your great contribution.

therealak12 commented:
> LGTM with one nit comment. I think we can land this today or tomorrow. :) Thanks again for all your great contribution.

Thanks for the review! I've addressed your comment - please let me know if you'd like any further adjustments.

wbpcode reviewed (approved):
LGTM. Thanks.
